### PR TITLE
Re-introduce distinction between CodeType and CodeTypeDispatch

### PR DESCRIPTION
**Pull request by mhammond**
This fixes a source of significant confusion in the code-oracle support.

On one hand, it has the concept of a `CodeType` which you get back from an Oracle via a `CodeDispatch` - so a code-type is closely tied to the oracle that supplied it. But:

* Every `CodeType` method takes a `CodeOracle` param - which is odd because the `CodeType` came from the oracle - it doesn't need the oracle to work.

* There's a blanket implementation of `CodeType` made available for `Type` - so not only do we now have a `CodeType` that didn't come from an Oracle, so clearly can't know the language semantics the Oracle abstracts away, there's now *two* implementations of `CodeType` for each type, with each behaving differently.

This cleans all that up:

* CodeType is only supplied by Oracles - there's no blanket implementation for Type.

* Therefore `CodeType` no longer takes Oracle's as args.

* Filters now take a `CodeTypeDispatch` - which *are* implemented for type, and explicitly use the oracle to get a `CodeType`

### CODE REVIEW

File: uniffi_bindgen/src/backend/types.rs [hunk 1]
```diff
@@ -65,46 +55,46 @@ pub trait CodeType {
...
+    fn initialization_fn(&self) -> Option<String> {
         None
     }
 }
 
 /// This trait is used to implement `CodeType` for `Type` and type-like structs (`Record`, `Enum`, `Field`,
 /// etc).  We forward all method calls to a `Box<dyn CodeType>`, which we get by calling
 /// `CodeOracle.find()`.
-pub trait CodeTypeDispatch {
+pub trait CodeTypeDispatch: Debug {
```
Comment (bendk) [inline uniffi_bindgen/src/backend/types.rs:97]: I would love to replace this trait with something like:

```
pub trait AsType {
   fn as_type(&self) -> Type;
}
```

The having filters that looked like:
```
pub fn type_name(t: &impl AsType) -> Result<String, askama::Error> {
    Ok(oracle.find(t.as_type()).type_label())
}
```

`AsType` is more general than `CodeTypeDispatch` and overall the code seems simpler to me.  Maybe this should be a separate PR though.

**Review by bendk — APPROVED** [review]
These changes look good to me.  Maybe we should consider the `AsType` suggestion here, but I'm also fine with merging this one and discussing that separately.

File: uniffi_bindgen/src/backend/types.rs (hunk 1 above)
Comment (mhammond) [reply uniffi_bindgen/src/backend/types.rs:97]: That's actually a great idea and I agree it's much cleaner - I put up #1588, let me know if that's not what you had in mind.

**Comment by mhammond** [comment]
Done in #1588
